### PR TITLE
[SPARK-12575][SQL] Grammar parity with existing SQL parser

### PR DESCRIPTION
**Pull request by hvanhovell**
In this PR the new CatalystQl parser stack reaches grammar parity with the old Parser-Combinator based SQL Parser. This PR also replaces all uses of the old Parser, and removes it from the code base.

Although the existing Hive and SQL parser dialects were mostly the same, some kinks had to be worked out:
- The SQL Parser allowed syntax like `APPROXIMATE(0.01) COUNT(DISTINCT a)`. In order to make this work we needed to hardcode approximate operators in the parser, or we would have to create an approximate expression. `APPROXIMATE_COUNT_DISTINCT(a, 0.01)` would also do the job and is much easier to maintain. So, this PR **removes** this keyword.
- The old SQL Parser supports `LIMIT` clauses in nested queries. This is **not supported** anymore. See https://github.com/apache/spark/pull/10689 for the rationale for this.
- Hive has a charset name char set literal combination it supports, for instance the following expression `_ISO-8859-1 0x4341464562616265` would yield this string: `CAFEbabe`. Hive will only allow charset names to start with an underscore. This is quite annoying in spark because as soon as you use a tuple names will start with an underscore. In this PR we **remove** this feature from the parser. It would be quite easy to implement such a feature as an Expression later on.
- Hive and the SQL Parser treat decimal literals differently. Hive will turn any decimal into a `Double` whereas the SQL Parser would convert a non-scientific decimal into a `BigDecimal`, and would turn a scientific decimal into a Double. We follow Hive's behavior here. The new parser supports a big decimal literal, for instance: `81923801.42BD`, which can be used when a big decimal is needed.

cc @rxin @viirya @marmbrus @yhuai @cloud-fan

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 1]
```diff
@@ -1063,7 +1063,7 @@ object functions extends LegacyFunctions {
    *
    * @group normal_funcs
    */
-  def expr(expr: String): Column = Column(SqlParser.parseExpression(expr))
+  def expr(expr: String): Column = Column(CatalystQl.parseExpression(expr))
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: I am not happy about this one: we are using an unconfigured parser here.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/SqlParserSuite.scala [hunk 2]
```diff
@@ -1,150 +0,0 @@
...
-import org.apache.spark.sql.catalyst.plans.PlanTest
-import org.apache.spark.sql.catalyst.plans.logical.{Command, LogicalPlan, OneRowRelation, Project}
-import org.apache.spark.unsafe.types.CalendarInterval
-
-private[sql] case class TestCommand(cmd: String) extends LogicalPlan with Command {
-  override def output: Seq[Attribute] = Seq.empty
-  override def children: Seq[LogicalPlan] = Seq.empty
-}
-
-private[sql] class SuperLongKeywordTestParser extends AbstractSparkSQLParser {
```
Comment (hvanhovell) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/SqlParserSuite.scala]: Allmost all tests have been move to CatalystQl suite in a previous PR.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala (hunk 1 above)
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Not pretty but: you could get the conf from the SQLContext getOrCreate / getActive?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/SqlParser.scala [hunk 3]
```diff
@@ -1,509 +0,0 @@
...
-import org.apache.spark.unsafe.types.CalendarInterval
-
-/**
- * A very simple SQL parser.  Based loosely on:
- * https://github.com/stephentu/scala-sql-parser/blob/master/src/main/scala/parser.scala
- *
- * Limitations:
- *  - Only supports a very limited subset of SQL.
- *
- * This is currently included mostly for illustrative purposes.  Users wanting more complete support
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/SqlParser.scala]: Good riddance!

**Comment by yhuai** [comment]
test this please

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystQlSuite.scala [hunk 4]
```diff
@@ -170,4 +171,10 @@ class CatalystQlSuite extends PlanTest {
     parser.parsePlan("select sum(product + 1) over (partition by (product + (1)) order by 2) " +
       "from windowData")
   }
+
+  test("string charset literal") {
+    compareExpressions(
+      parser.parseExpression("'ISO-8859-1' 0x4341464562616265"),
```
Comment (rxin) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystQlSuite.scala]: are there any other databases that allow this feature?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSQLParser.scala [hunk 5]
```diff
@@ -31,7 +31,12 @@ import org.apache.spark.sql.types.StringType
  *
  * @param fallback A function that parses an input string to a logical plan
  */
-class SparkSQLParser(fallback: String => LogicalPlan) extends AbstractSparkSQLParser {
+class SparkSQLParser(fallback: => ParserDialect) extends AbstractSparkSQLParser {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSQLParser.scala]: should this just be a field now rather than call by name?

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala (hunk 1 above)
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: If we do that, we should make sure we have a default one when there is no active context.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystQl.scala [hunk 6]
```diff
@@ -787,76 +783,92 @@ https://cwiki.apache.org/confluence/display/Hive/Enhanced+Aggregation%2C+Cube%2C
...
+      text match {
+        case INTEGRAL() =>
+          BigDecimal(text) match {
+            case v if v.isValidInt =>
+              Literal(v.intValue())
+            case v if v.isValidLong =>
+              Literal(v.longValue())
+            case v => Literal(v.underlying())
+          }
+        case DECIMAL(_*) =>
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystQl.scala]: I think we should make them consistent in 2.0.

File: sql/catalyst/src/main/antlr3/org/apache/spark/sql/catalyst/parser/FromClauseParser.g [hunk 7]
```diff
@@ -206,11 +206,8 @@ tableName
 @init { gParent.pushMsg("table name", state); }
 @after { gParent.popMsg(state); }
     :
-    db=identifier DOT tab=identifier
-    -> ^(TOK_TABNAME $db $tab)
-    |
-    tab=identifier
-    -> ^(TOK_TABNAME $tab)
+    id1=identifier (DOT id2=identifier)?
```
Comment (rxin) [inline sql/catalyst/src/main/antlr3/org/apache/spark/sql/catalyst/parser/FromClauseParser.g]: what is this change?

**Comment by rxin** [comment]
Looks pretty good overall.

**Comment by hvanhovell** [comment]
retest this please

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystQlSuite.scala (hunk 4 above)
Comment (hvanhovell) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystQlSuite.scala]: MySQL apparently supports this: https://docs.oracle.com/cd/E17952_01/refman-5.5-en/charset-literal.html

File: sql/catalyst/src/main/antlr3/org/apache/spark/sql/catalyst/parser/FromClauseParser.g (hunk 7 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/antlr3/org/apache/spark/sql/catalyst/parser/FromClauseParser.g]: They are semantically equal. The old one is a bit longer. That is all.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala (hunk 1 above)
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Could we also move this to ```SQLImplicits```?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystQl.scala (hunk 6 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystQl.scala]: I agree. Which default should we pick then? The parser currently supports a BigDecimal literal, so we could go with that, and default to doubles when a decimal is encountered.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSQLParser.scala (hunk 5 above)
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSQLParser.scala]: I thought the same thing. The problem is that we we pass the ```SQLContext.getDialect()``` method to the parser (same as before): this allows us to change the dialect whenever we want.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala (hunk 1 above)
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: That won't work for Java though.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSQLParser.scala (hunk 5 above)
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSQLParser.scala]: OK in that case we should have a comment here explaining the choice.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala (hunk 1 above)
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Yeah, totally forgot about that.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystQlSuite.scala (hunk 4 above)
Comment (rxin) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystQlSuite.scala]: Talked to a few more people about this. I think it's best to just drop this. Only MySQL and Hive support this, and we cannot support the identical syntax anyway. I'd say if this is a really desired feature, we can just build a function for it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystQl.scala (hunk 6 above)
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystQl.scala]: as discussed offline, converting it to double makes sense.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystQlSuite.scala (hunk 4 above)
Comment (hvanhovell) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystQlSuite.scala]: I'll drop the feature.

File: sql/catalyst/src/main/antlr3/org/apache/spark/sql/catalyst/parser/ExpressionParser.g [hunk 8]
```diff
@@ -163,22 +163,38 @@ timestampLiteral
...
+      -> ^(TOK_INTERVAL
+          ^(TOK_INTERVAL_YEAR_LITERAL $year?)
+          ^(TOK_INTERVAL_MONTH_LITERAL $month?)
+          ^(TOK_INTERVAL_WEEK_LITERAL $week?)
+          ^(TOK_INTERVAL_DAY_LITERAL $day?)
+          ^(TOK_INTERVAL_HOUR_LITERAL $hour?)
+          ^(TOK_INTERVAL_MINUTE_LITERAL $minute?)
+          ^(TOK_INTERVAL_SECOND_LITERAL $second?)
+          ^(TOK_INTERVAL_MILLISECOND_LITERAL $millisecond?)
+          ^(TOK_INTERVAL_MICROSECOND_LITERAL $microsecond?))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/antlr3/org/apache/spark/sql/catalyst/parser/ExpressionParser.g]: are these copied from hive?

Comment (hvanhovell) [inline sql/catalyst/src/main/antlr3/org/apache/spark/sql/catalyst/parser/ExpressionParser.g]: No this is what was supported in the old SqlParser.

Comment (cloud-fan) [inline sql/catalyst/src/main/antlr3/org/apache/spark/sql/catalyst/parser/ExpressionParser.g]: are you trying to support both hive's and our internal literal grammar?

Comment (hvanhovell) [inline sql/catalyst/src/main/antlr3/org/apache/spark/sql/catalyst/parser/ExpressionParser.g]: Hive does not support the following interval: ```1 year 3 month 10 milliseconds```

Comment (hvanhovell) [inline sql/catalyst/src/main/antlr3/org/apache/spark/sql/catalyst/parser/ExpressionParser.g]: are you trying to support both hive's and our interval literal grammar?

Not in all cases. I try to when I can. I think in this case it adds value.

Comment (rxin) [inline sql/catalyst/src/main/antlr3/org/apache/spark/sql/catalyst/parser/ExpressionParser.g]: +1 on supporting both. actually we have to here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystQl.scala [hunk 9]
```diff
@@ -787,76 +783,92 @@ https://cwiki.apache.org/confluence/display/Hive/Enhanced+Aggregation%2C+Cube%2C
...
+      Literal(ParseUtils.charSetString(unquoteString(ast.children.head.text), ast.children(1).text))
+
+    case ast if ast.tokenType == SparkSqlParser.TOK_INTERVAL_YEAR_MONTH_LITERAL =>
+      Literal(CalendarInterval.fromYearMonthString(ast.children.head.text))
+
+    case ast if ast.tokenType == SparkSqlParser.TOK_INTERVAL_DAY_TIME_LITERAL =>
+      Literal(CalendarInterval.fromDayTimeString(ast.children.head.text))
+
+    case Token("TOK_INTERVAL", elements) =>
+      val (interval, updated) = elements.foldLeft((new CalendarInterval(0, 0), false)) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystQl.scala]: How about we check `elements.isEmpty` first and throw exception if needed, and then `foldLeft`? then we don't need this `updated` variable.

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystQl.scala]: actually instead of foldLeft with two values, it might be easier and more clear to write this as a loop and just mutate two variables.

**Comment by hvanhovell** [comment]
retest this please

**Comment by hvanhovell** [comment]
getting weird seemingly unrelated python error.

**Comment by rxin** [comment]
LGTM. Will merge once tests pass.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlus.scala [hunk 10]
```diff
@@ -135,7 +135,7 @@ case class HyperLogLogPlusPlus(
 
   override def dataType: DataType = LongType
 
-  override def inputTypes: Seq[AbstractDataType] = Seq(AnyDataType)
+  override def inputTypes: Seq[AbstractDataType] = Seq(AnyDataType, DoubleType)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlus.scala]: why this change?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlus.scala [hunk 11]
```diff
@@ -447,6 +447,7 @@ object HyperLogLogPlusPlus {
 
   private def validateDoubleLiteral(exp: Expression): Double = exp match {
     case Literal(d: Double, DoubleType) => d
+    case Literal(dec: Decimal, dt: DecimalType) => dec.toDouble
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlus.scala]: is it always safe to call `toDouble`?

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlus.scala]: I changed this because the old SQL parser was passing Decimal to HyperLogLogPlusPlus.

The pattern match eliminates the possibility of a NPE. Spark ```Decimal``` delegates the ```toDouble``` call to java ```BigDecimal``` which will return ```(-)Infinity``` if the decimal is outside of the double range. We should be good here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlus.scala (hunk 10 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlus.scala]: HyperLogLogPlusPlus was failing because I was passing it Decimal literals. I thought I could solve this by casting. While this is not relevant anymore, I still think this is valid.

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlus.scala]: `inputTypes` is used to check input types, however, `HyperLogLogPlusPlus` only have one input, see https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlus.scala#L132. 

So we don't need to give 2 type constraints here.

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlus.scala]: Yeah you are right. I am fixing it now.

**Comment by cloud-fan** [comment]
LGTM

**Comment by rxin** [comment]
Thanks - going to merge this.
